# Advent of Code, 4th Dec 2024

## Ceres Search - Part 1

"Looks like the Chief's not here. Next!" One of The Historians pulls out a device and pushes the only button on it. After a brief flash, you recognize the interior of the Ceres monitoring station!

As the search for the Chief continues, a small Elf who lives on the station tugs on your shirt; she'd like to know if you could help her with her **word search** (your puzzle input). She only has to find one word: XMAS.

This word search allows words to be horizontal, vertical, diagonal, written backwards, or even overlapping other words. It's a little unusual, though, as you don't merely need to find one instance of XMAS - you need to find **all of them**. Here are a few ways XMAS might appear, where irrelevant characters have been replaced with .:

```
..X...
.SAMX.
.A..A.
XMAS.S
.X....
```

The actual word search will be full of letters instead. For example:

```
MMMSXXMASM
MSAMXMSMSA
AMXSXMAAMM
MSAMASMSMX
XMASAMXAMM
XXAMMXXAMA
SMSMSASXSS
SAXAMASAAA
MAMMMXMMMM
MXMXAXMASX
```

In this word search, XMAS occurs a total of 18 times; here's the same word search again, but where letters not involved in any XMAS have been replaced with .:

```
....XXMAS.
.SAMXMS...
...S..A...
..A.A.MS.X
XMASAMX.MM
X.....XA.A
S.S.S.S.SS
.A.A.A.A.A
..M.M.M.MM
.X.X.XMASX
```

Take a look at the little Elf's word search. **How many times does XMAS appear?**

### Read in the data

In [1]:
#!F#
let inputFilePath = "aoc-2024-12-04-puzzle-input.txt"

Read the lines from the file.

In [2]:
open System.IO

let lines: string list = File.ReadAllLines(inputFilePath) |> Array.toList
lines.Length

140

In [3]:
lines |> List.map (fun str -> str.Length) |> Set.ofList

[ 140 ] Count 1 (values) [ 140 ]

So we see that there are 140 lines each of 140 characters, as we might have expected.

### Solve the problem

First, lets search for horizontal occurrences.

In [4]:
let searchStr = "XMAS"

let reverseString (input: string) =
    input |> Seq.rev |> Seq.toArray |> System.String

let reverseStr = searchStr |> reverseString

In [5]:
let countOccurrences (searchStr: string) (str: string): int =
    let rec innerCount (searchStr: string) (total: int) (str: string): int =
        match str with
        | "" -> total
        | matchedStr when matchedStr.StartsWith(searchStr) -> innerCount searchStr (total + 1) str.[1..]
        | shortStr when str.Length < searchStr.Length -> total
        | other -> innerCount searchStr total str.[1..]
    innerCount searchStr 0 str

Let's check that this can find overlapping occurrences.

In [6]:
countOccurrences "aba" "abababa"

3

Now let's do the actual search for horizontal occurrences.

In [7]:
let horizontalOccurrences =
    (lines |> List.map (fun line -> countOccurrences searchStr line) |> List.sum) +
    (lines |> List.map (fun line -> countOccurrences reverseStr line) |> List.sum)
horizontalOccurrences

416

To do vertical occurrences, we transpose the lines.

In [11]:
let rec transposeLines (lines: string list): string list = // assumption is that all lines in the list have the same length
    let transposedLineSeq = seq {
        let verticalHead: string = seq { for line in lines do yield line.[0] } |> Seq.toArray |> System.String
        yield verticalHead
        if (lines.[0].Length > 1)
        then
            let verticalTail: string list = seq { for line in lines do yield line.[1..] } |> Seq.toList
            yield! (transposeLines verticalTail)
    }
    transposedLineSeq |> Seq.toList

In [12]:
transposeLines ["abc"; "def"; "ghi"]

[ adg, beh, cfi ] HeadOrDefault adg TailOrNull [ beh, cfi ] HeadOrDefault beh TailOrNull [ cfi ] HeadOrDefault cfi TailOrNull [ ] HeadOrDefault <null> TailOrNull <null> Head System.InvalidOperationException: The input list was empty.\r\n at Microsoft.FSharp.Collections.FSharpList`1.get_Head() in D:\a\_work\1\s\src\FSharp.Core\prim-types.fs:line 4142\r\n at lambda_method113(Closure, FSharpList`1)\r\n at Microsoft.DotNet.Interactive.Formatting.MemberAccessor`1.GetVa... TargetSite T get_Head() Name get_Head DeclaringType Microsoft.FSharp.Collections.FSharpList`1[T] ReflectedType Microsoft.FSharp.Collections.FSharpList`1[T] MemberType Method MetadataToken 100663765 Module FSharp.Core.dll IsSecurityCritical True IsSecuritySafeCritical False IsSecurityTransparent False MethodHandle System.RuntimeMethodHandle Attributes Public, HideBySig, SpecialName CallingConvention Standard, HasThis ReturnType T ReturnTypeCustomAttributes T ReturnParameter T IsCollectible False IsGenericMethod False IsGenericMethodDefinition False ContainsGenericParameters True MethodImplementationFlags IL IsAbstract False IsConstructor False IsFinal False IsHideBySig True IsSpecialName True IsStatic False IsVirtual False IsAssembly False IsFamily False IsFamilyAndAssembly False IsFamilyOrAssembly False IsPrivate False IsPublic True IsConstructedGenericMethod False CustomAttributes [ ] Message The input list was empty. Data (empty) InnerException <null> HelpLink <null> Source FSharp.Core HResult -2146233079 StackTrace at Microsoft.FSharp.Collections.FSharpList`1.get_Head() in D:\a\_work\1\s\src\FSharp.Core\prim-types.fs:line 4142
 at lambda_method113(Closure, FSharpList`1)
 at Microsoft.DotNet.Interactive.Formatting.MemberAccessor`1.GetValueOrException(T instance) in D:\a\_work\1\s\src\Microsoft.DotNet.Interactive.Formatting\MemberAccessor{T}.cs:line 58 Tail System.InvalidOperationException: The input list was empty.\r\n at Microsoft.FSharp.Collections.FSharpList`1.get_Tail() in D:\a\_work\1\s\src\FSharp.Core\prim-types.fs:line 4147\r\n at lambda_method114(Closure, FSharpList`1)\r\n at Microsoft.DotNet.Interactive.Formatting.MemberAccessor`1.GetVa... TargetSite Microsoft.FSharp.Collections.FSharpList`1[T] get_Tail() Name get_Tail DeclaringType Microsoft.FSharp.Collections.FSharpList`1[T] ReflectedType Microsoft.FSharp.Collections.FSharpList`1[T] MemberType Method MetadataToken 100663766 Module FSharp.Core.dll IsSecurityCritical True IsSecuritySafeCritical False IsSecurityTransparent False MethodHandle System.RuntimeMethodHandle Attributes Public, HideBySig, SpecialName CallingConvention Standard, HasThis ReturnType Microsoft.FSharp.Collections.FSharpList`1[T] ReturnTypeCustomAttributes Microsoft.FSharp.Collections.FSharpList`1[T] ReturnParameter Microsoft.FSharp.Collections.FSharpList`1[T] IsCollectible False IsGenericMethod False IsGenericMethodDefinition False ContainsGenericParameters True MethodImplementationFlags IL IsAbstract False IsConstructor False IsFinal False IsHideBySig True IsSpecialName True IsStatic False IsVirtual False IsAssembly False IsFamily False IsFamilyAndAssembly False IsFamilyOrAssembly False IsPrivate False IsPublic True IsConstructedGenericMethod False CustomAttributes [ ] Message The input list was empty. Data (empty) InnerException <null> HelpLink <null> Source FSharp.Core HResult -2146233079 StackTrace at Microsoft.FSharp.Collections.FSharpList`1.get_Tail() in D:\a\_work\1\s\src\FSharp.Core\prim-types.fs:line 4147
 at lambda_method114(Closure, FSharpList`1)
 at Microsoft.DotNet.Interactive.Formatting.MemberAccessor`1.GetValueOrException(T instance) in D:\a\_work\1\s\src\Microsoft.DotNet.Interactive.Formatting\MemberAccessor{T}.cs:line 58 (values) [ ] Head cfi Tail [ ] HeadOrDefault <null> TailOrNull <null> Head System.InvalidOperationException: The input list was empty.\r\n at Microsoft.FSharp.Collections.FSharpList`1.get_Head() in D:\a\_work\1\s\src\FSharp.Core\prim-types.fs:line 4142\r\n at lambda_method113(Closure, FSharpList`

In [13]:
let transposedLines = lines |> transposeLines

let verticalOccurrences =
    (transposedLines |> List.map (fun line -> countOccurrences searchStr line) |> List.sum) +
    (transposedLines |> List.map (fun line -> countOccurrences reverseStr line) |> List.sum)
verticalOccurrences

412

Now, to do the diagonal occurrences, we do a "diagonal transposition" of the lines.  This can be done in either of two directions.  Let's do left diagonals first.

In [36]:
let transposeLeftDiagonalLines (lines: string list): string list =
    // For N lines, there will be 2*N-1 diagonals
    let innerTake (depth: int) (lines: string list): string * (string list) = // returns the first char from each of the first 'depth' lines, along with the unused tails
        let headSeq = seq {
            for idx in 0..(depth-1) do
                if (lines.[idx].Length > 0)
                then yield lines.[idx].[0]
        }
        let head = headSeq |> Seq.toArray |> System.String
        let tailSeq = seq {
            for idx in 0..(depth-1) do
                if (lines.[idx].Length > 1)
                then yield lines.[idx].[1..]
            for idx in depth..(lines.Length-1) do
                if (lines.[idx].Length > 0)
                then yield lines.[idx]
        }
        let tailList = tailSeq |> Seq.toList
        (head, tailList)
    let rec iterate (diagonalsAndLines: string list * string list) (depth: int) =
        let diagonals, lines = diagonalsAndLines
        let newDiagonal, newLines = innerTake depth lines
        let newDiagonals = List.concat [diagonals; [newDiagonal]]
        (newDiagonals, newLines)
    let depthSeries = List.concat [[1..(lines.Length)]; [(lines.Length-1)..(-1)..1]] // 1 to N to 1 again
    let diagonals, remainingLines = depthSeries |> List.fold iterate ([], lines)
    diagonals
    

In [37]:
transposeLeftDiagonalLines ["abc"; "def"; "ghi"]

[ a, bd, ceg, fh, i ] HeadOrDefault a TailOrNull [ bd, ceg, fh, i ] HeadOrDefault bd TailOrNull [ ceg, fh, i ] HeadOrDefault ceg TailOrNull [ fh, i ] HeadOrDefault fh TailOrNull [ i ] HeadOrDefault i TailOrNull [ ] HeadOrDefault <null> TailOrNull <null> Head System.InvalidOperationException: The input list was empty.
 at Microsoft.FSharp.Collections.FSharpList`1.get_Head() in D:\a\_work\1\s\src\FSharp.Core\prim-types.fs:line 4142
 at lambda_method113(Closure, FSharpList`1)
 at Microsoft.DotNet.Interactive.Formatting.MemberAccessor`1.GetValueOrException(T instance) in D:\a\_work\1\s\src\Microsoft.DotNet.Interactive.Formatting\MemberAccessor{T}.cs:line 58 Tail System.InvalidOperationException: The input list was empty.
 at Microsoft.FSharp.Collections.FSharpList`1.get_Tail() in D:\a\_work\1\s\src\FSharp.Core\prim-types.fs:line 4147
 at lambda_method114(Closure, FSharpList`1)
 at Microsoft.DotNet.Interactive.Formatting.MemberAccessor`1.GetValueOrException(T instance) in D:\a\_work\1\s\src\Microsoft.DotNet.Interactive.Formatting\MemberAccessor{T}.cs:line 58 (values) [ ] Head i Tail [ ] HeadOrDefault <null> TailOrNull <null> Head System.InvalidOperationException: The input list was empty.
 at Microsoft.FSharp.Collections.FSharpList`1.get_Head() in D:\a\_work\1\s\src\FSharp.Core\prim-types.fs:line 4142
 at lambda_method113(Closure, FSharpList`1)
 at Microsoft.DotNet.Interactive.Formatting.MemberAccessor`1.GetValueOrException(T instance) in D:\a\_work\1\s\src\Microsoft.DotNet.Interactive.Formatting\MemberAccessor{T}.cs:line 58 Tail System.InvalidOperationException: The input list was empty.
 at Microsoft.FSharp.Collections.FSharpList`1.get_Tail() in D:\a\_work\1\s\src\FSharp.Core\prim-types.fs:line 4147
 at lambda_method114(Closure, FSharpList`1)
 at Microsoft.DotNet.Interactive.Formatting.MemberAccessor`1.GetValueOrException(T instance) in D:\a\_work\1\s\src\Microsoft.DotNet.Interactive.Formatting\MemberAccessor{T}.cs:line 58 (values) [ ] (values) [ i ] Head fh Tail [ i ] HeadOrDefault i TailOrNull [ ] HeadOrDefault <null> TailOrNull <null> Head System.InvalidOperationException: The input list was empty.
 at Microsoft.FSharp.Collections.FSharpList`1.get_Head() in D:\a\_work\1\s\src\FSharp.Core\prim-types.fs:line 4142
 at lambda_method113(Closure, FSharpList`1)
 at Microsoft.DotNet.Interactive.Formatting.MemberAccessor`1.GetValueOrException(T instance) in D:\a\_work\1\s\src\Microsoft.DotNet.Interactive.Formatting\MemberAccessor{T}.cs:line 58 Tail System.InvalidOperationException: The input list was empty.
 at Microsoft.FSharp.Collections.FSharpList`1.get_Tail() in D:\a\_work\1\s\src\FSharp.Core\prim-types.fs:line 4147
 at lambda_method114(Closure, FSharpList`1)
 at Microsoft.DotNet.Interactive.Formatting.MemberAccessor`1.GetValueOrException(T instance) in D:\a\_work\1\s\src\Microsoft.DotNet.Interactive.Formatting\MemberAccessor{T}.cs:line 58 (values) [ ] Head i Tail [ ] HeadOrDefault <null> TailOrNull <null> Head System.InvalidOperationException: The input list was empty.
 at Microsoft.FSharp.Collections.FSharpList`1.get_Head() in D:\a\_work\1\s\src\FSharp.Core\prim-types.fs:line 4142
 at lambda_method113(Closure, FSharpList`1)
 at Microsoft.DotNet.Interactive.Formatting.MemberAccessor`1.GetValueOrException(T instance) in D:\a\_work\1\s\src\Microsoft.DotNet.Interactive.Formatting\MemberAccessor{T}.cs:line 58 Tail System.InvalidOperationException: The input list was empty.
 at Microsoft.FSharp.Collections.FSharpList`1.get_Tail() in D:\a\_work\1\s\src\FSharp.Core\prim-types.fs:line 4147
 at lambda_method114(Closure, FSharpList`1)
 at Microsoft.DotNet.Interactive.Formatting.MemberAccessor`1.GetValueOrException(T instance) in D:\a\_work\1\s\src\Microsoft.DotNet.Interactive.Formatting\MemberAccessor{T}.cs:line 58 (values) [ ] (values) [ i ] (values) [ fh, i ] Head ceg Tail [ fh, i ] HeadOrDefault fh TailOrNull [ i ] HeadOrDefault i TailOrNull [ ] HeadOrDefault <null> TailOrNull <null> Head System.InvalidOperationEx

In [38]:
let transposedLeftDiagonalLines = lines |> transposeLeftDiagonalLines

let transposedLeftDiagonalOccurrences =
    (transposedLeftDiagonalLines |> List.map (fun line -> countOccurrences searchStr line) |> List.sum) +
    (transposedLeftDiagonalLines |> List.map (fun line -> countOccurrences reverseStr line) |> List.sum)
transposedLeftDiagonalOccurrences

815

To do right diagonals, we just reverse the strings and then do their left diagonals.

In [44]:
let transposeRightDiagonalLines (lines: string list): string list =
    let revLines = lines |> List.map reverseString
    transposeLeftDiagonalLines revLines

In [45]:
transposeRightDiagonalLines ["abc"; "def"; "ghi"]

[ c, bf, aei, dh, g ] HeadOrDefault c TailOrNull [ bf, aei, dh, g ] HeadOrDefault bf TailOrNull [ aei, dh, g ] HeadOrDefault aei TailOrNull [ dh, g ] HeadOrDefault dh TailOrNull [ g ] HeadOrDefault g TailOrNull [ ] HeadOrDefault <null> TailOrNull <null> Head System.InvalidOperationException: The input list was empty.
 at Microsoft.FSharp.Collections.FSharpList`1.get_Head() in D:\a\_work\1\s\src\FSharp.Core\prim-types.fs:line 4142
 at lambda_method113(Closure, FSharpList`1)
 at Microsoft.DotNet.Interactive.Formatting.MemberAccessor`1.GetValueOrException(T instance) in D:\a\_work\1\s\src\Microsoft.DotNet.Interactive.Formatting\MemberAccessor{T}.cs:line 58 Tail System.InvalidOperationException: The input list was empty.
 at Microsoft.FSharp.Collections.FSharpList`1.get_Tail() in D:\a\_work\1\s\src\FSharp.Core\prim-types.fs:line 4147
 at lambda_method114(Closure, FSharpList`1)
 at Microsoft.DotNet.Interactive.Formatting.MemberAccessor`1.GetValueOrException(T instance) in D:\a\_work\1\s\src\Microsoft.DotNet.Interactive.Formatting\MemberAccessor{T}.cs:line 58 (values) [ ] Head g Tail [ ] HeadOrDefault <null> TailOrNull <null> Head System.InvalidOperationException: The input list was empty.
 at Microsoft.FSharp.Collections.FSharpList`1.get_Head() in D:\a\_work\1\s\src\FSharp.Core\prim-types.fs:line 4142
 at lambda_method113(Closure, FSharpList`1)
 at Microsoft.DotNet.Interactive.Formatting.MemberAccessor`1.GetValueOrException(T instance) in D:\a\_work\1\s\src\Microsoft.DotNet.Interactive.Formatting\MemberAccessor{T}.cs:line 58 Tail System.InvalidOperationException: The input list was empty.
 at Microsoft.FSharp.Collections.FSharpList`1.get_Tail() in D:\a\_work\1\s\src\FSharp.Core\prim-types.fs:line 4147
 at lambda_method114(Closure, FSharpList`1)
 at Microsoft.DotNet.Interactive.Formatting.MemberAccessor`1.GetValueOrException(T instance) in D:\a\_work\1\s\src\Microsoft.DotNet.Interactive.Formatting\MemberAccessor{T}.cs:line 58 (values) [ ] (values) [ g ] Head dh Tail [ g ] HeadOrDefault g TailOrNull [ ] HeadOrDefault <null> TailOrNull <null> Head System.InvalidOperationException: The input list was empty.
 at Microsoft.FSharp.Collections.FSharpList`1.get_Head() in D:\a\_work\1\s\src\FSharp.Core\prim-types.fs:line 4142
 at lambda_method113(Closure, FSharpList`1)
 at Microsoft.DotNet.Interactive.Formatting.MemberAccessor`1.GetValueOrException(T instance) in D:\a\_work\1\s\src\Microsoft.DotNet.Interactive.Formatting\MemberAccessor{T}.cs:line 58 Tail System.InvalidOperationException: The input list was empty.
 at Microsoft.FSharp.Collections.FSharpList`1.get_Tail() in D:\a\_work\1\s\src\FSharp.Core\prim-types.fs:line 4147
 at lambda_method114(Closure, FSharpList`1)
 at Microsoft.DotNet.Interactive.Formatting.MemberAccessor`1.GetValueOrException(T instance) in D:\a\_work\1\s\src\Microsoft.DotNet.Interactive.Formatting\MemberAccessor{T}.cs:line 58 (values) [ ] Head g Tail [ ] HeadOrDefault <null> TailOrNull <null> Head System.InvalidOperationException: The input list was empty.
 at Microsoft.FSharp.Collections.FSharpList`1.get_Head() in D:\a\_work\1\s\src\FSharp.Core\prim-types.fs:line 4142
 at lambda_method113(Closure, FSharpList`1)
 at Microsoft.DotNet.Interactive.Formatting.MemberAccessor`1.GetValueOrException(T instance) in D:\a\_work\1\s\src\Microsoft.DotNet.Interactive.Formatting\MemberAccessor{T}.cs:line 58 Tail System.InvalidOperationException: The input list was empty.
 at Microsoft.FSharp.Collections.FSharpList`1.get_Tail() in D:\a\_work\1\s\src\FSharp.Core\prim-types.fs:line 4147
 at lambda_method114(Closure, FSharpList`1)
 at Microsoft.DotNet.Interactive.Formatting.MemberAccessor`1.GetValueOrException(T instance) in D:\a\_work\1\s\src\Microsoft.DotNet.Interactive.Formatting\MemberAccessor{T}.cs:line 58 (values) [ ] (values) [ g ] (values) [ dh, g ] Head aei Tail [ dh, g ] HeadOrDefault dh TailOrNull [ g ] HeadOrDefault g TailOrNull [ ] HeadOrDefault <null> TailOrNull <null> Head System.InvalidOperationEx

In [46]:
let transposedRightDiagonalLines = lines |> transposeRightDiagonalLines

let transposedRightDiagonalOccurrences =
    (transposedRightDiagonalLines |> List.map (fun line -> countOccurrences searchStr line) |> List.sum) +
    (transposedRightDiagonalLines |> List.map (fun line -> countOccurrences reverseStr line) |> List.sum)
transposedRightDiagonalOccurrences

758

In [49]:
let totalOccurrences = horizontalOccurrences + verticalOccurrences + transposedLeftDiagonalOccurrences + transposedRightDiagonalOccurrences
totalOccurrences

2401

## Ceres Search - Part 2

The Elf looks quizzically at you. Did you misunderstand the assignment?

Looking for the instructions, you flip over the word search to find that this isn't actually an **XMAS** puzzle; it's an **X-MAS** puzzle in which you're supposed to find two MAS in the shape of an X. One way to achieve that is like this:

```
M.S
.A.
M.S
```

Irrelevant characters have again been replaced with . in the above diagram. Within the X, each MAS can be written forwards or backwards.

Here's the same example from before, but this time all of the X-MASes have been kept instead:

```
.M.S......
..A..MSMS.
.M.S.MAA..
..A.ASMSM.
.M.S.M....
..........
S.S.S.S.S.
.A.A.A.A..
M.M.M.M.M.
..........
```

In this example, an X-MAS appears **9** times.

Flip the word search from the instructions back over to the word search side and try again. **How many times does an X-MAS appear?**

In [50]:
let extractRectangles (width: int) (height: int) (lines: string list): string list list = // strings must all be the same length
    let rowCount = lines.Length
    let colCount = lines.[0].Length
    let rectSeq = seq {
        for startCol in 0..(colCount-width) do
            for startRow in 0..(rowCount-height) do
                let rectRowSeq = seq {
                    for row in startRow..(startRow+height-1) do
                        yield lines.[row].[startCol..(startCol+width-1)]
                }
                yield rectRowSeq |> Seq.toList
    }
    rectSeq |> Seq.toList

In [51]:
extractRectangles 3 2 ["abcde" ; "fghij" ; "klmno"]

[ [ abc, fgh ], [ fgh, klm ], [ bcd, ghi ], [ ghi, lmn ], [ cde, hij ], [ hij, mno ] ] HeadOrDefault [ abc, fgh ] HeadOrDefault abc TailOrNull [ fgh ] HeadOrDefault fgh TailOrNull [ ] HeadOrDefault <null> TailOrNull <null> Head System.InvalidOperationException: The input list was empty.\r\n at Microsoft.FSharp.Collections.FSharpList`1.get_Head() in D:\a\_work\1\s\src\FSharp.Core\prim-types.fs:line 4142\r\n at lambda_method113(Closure, FSharpList`1)\r\n at Microsoft.DotNet.Interactive.Formatting.MemberAccessor`1.GetVa... TargetSite T get_Head() Name get_Head DeclaringType Microsoft.FSharp.Collections.FSharpList`1[T] ReflectedType Microsoft.FSharp.Collections.FSharpList`1[T] MemberType Method MetadataToken 100663765 Module FSharp.Core.dll IsSecurityCritical True IsSecuritySafeCritical False IsSecurityTransparent False MethodHandle System.RuntimeMethodHandle Attributes Public, HideBySig, SpecialName CallingConvention Standard, HasThis ReturnType T ReturnTypeCustomAttributes T ReturnParameter T IsCollectible False IsGenericMethod False IsGenericMethodDefinition False ContainsGenericParameters True MethodImplementationFlags IL IsAbstract False IsConstructor False IsFinal False IsHideBySig True IsSpecialName True IsStatic False IsVirtual False IsAssembly False IsFamily False IsFamilyAndAssembly False IsFamilyOrAssembly False IsPrivate False IsPublic True IsConstructedGenericMethod False CustomAttributes [ ] Message The input list was empty. Data (empty) InnerException <null> HelpLink <null> Source FSharp.Core HResult -2146233079 StackTrace at Microsoft.FSharp.Collections.FSharpList`1.get_Head() in D:\a\_work\1\s\src\FSharp.Core\prim-types.fs:line 4142
 at lambda_method113(Closure, FSharpList`1)
 at Microsoft.DotNet.Interactive.Formatting.MemberAccessor`1.GetValueOrException(T instance) in D:\a\_work\1\s\src\Microsoft.DotNet.Interactive.Formatting\MemberAccessor{T}.cs:line 58 Tail System.InvalidOperationException: The input list was empty.\r\n at Microsoft.FSharp.Collections.FSharpList`1.get_Tail() in D:\a\_work\1\s\src\FSharp.Core\prim-types.fs:line 4147\r\n at lambda_method114(Closure, FSharpList`1)\r\n at Microsoft.DotNet.Interactive.Formatting.MemberAccessor`1.GetVa... TargetSite Microsoft.FSharp.Collections.FSharpList`1[T] get_Tail() Name get_Tail DeclaringType Microsoft.FSharp.Collections.FSharpList`1[T] ReflectedType Microsoft.FSharp.Collections.FSharpList`1[T] MemberType Method MetadataToken 100663766 Module FSharp.Core.dll IsSecurityCritical True IsSecuritySafeCritical False IsSecurityTransparent False MethodHandle System.RuntimeMethodHandle Attributes Public, HideBySig, SpecialName CallingConvention Standard, HasThis ReturnType Microsoft.FSharp.Collections.FSharpList`1[T] ReturnTypeCustomAttributes Microsoft.FSharp.Collections.FSharpList`1[T] ReturnParameter Microsoft.FSharp.Collections.FSharpList`1[T] IsCollectible False IsGenericMethod False IsGenericMethodDefinition False ContainsGenericParameters True MethodImplementationFlags IL IsAbstract False IsConstructor False IsFinal False IsHideBySig True IsSpecialName True IsStatic False IsVirtual False IsAssembly False IsFamily False IsFamilyAndAssembly False IsFamilyOrAssembly False IsPrivate False IsPublic True IsConstructedGenericMethod False CustomAttributes [ ] Message The input list was empty. Data (empty) InnerException <null> HelpLink <null> Source FSharp.Core HResult -2146233079 StackTrace at Microsoft.FSharp.Collections.FSharpList`1.get_Tail() in D:\a\_work\1\s\src\FSharp.Core\prim-types.fs:line 4147
 at lambda_method114(Closure, FSharpList`1)
 at Microsoft.DotNet.Interactive.Formatting.MemberAccessor`1.GetValueOrException(T instance) in D:\a\_work\1\s\src\Microsoft.DotNet.Interactive.Formatting\MemberAccessor{T}.cs:line 58 (values) [ ] Head fgh Tail [ ] HeadOrDefault <null> TailOrNull <null> Head System.InvalidOperationException: The input list was empty.\r\n at Microsoft.FSharp.Collections.FSharpList`1.get_Head() in D:\a\_work\1\s\src\FSharp.Core\prim-types.fs:

In [52]:
let masCandidateRectangles = lines |> extractRectangles 3 3
masCandidateRectangles.Length

19044

In [60]:
let x_masCount (candidateRectangle: string list): int = // candidateRectangle must be a list of 3 strings, each 3 characters long
    if (candidateRectangle.[1].[1] <> 'A')
    then 0
    else
        let masChecks = [
            if (candidateRectangle.[0].[0] = 'M' && candidateRectangle.[2].[2] = 'S') then 1 else 0 ;
            if (candidateRectangle.[2].[2] = 'M' && candidateRectangle.[0].[0] = 'S') then 1 else 0 ;
            if (candidateRectangle.[0].[2] = 'M' && candidateRectangle.[2].[0] = 'S') then 1 else 0 ;
            if (candidateRectangle.[2].[0] = 'M' && candidateRectangle.[0].[2] = 'S') then 1 else 0
        ]
        let masCheckSum = masChecks |> List.sum
        if (masCheckSum = 2) then 1 else 0 // two "MAS" diagonals = one X-MAS

In [61]:
x_masCount ["MXS" ; "XAX" ; "MXS"] // result should be 2

1

So let's count how many X-MAS occurrences we have.

In [62]:
masCandidateRectangles |> List.map x_masCount |> List.sum

1822